### ISSUE TITLE
Stack lost for errors thrown after model is accessed -  WARNING: Something seems to be wrong with this function.

### ISSUE THREAD

**Issue by wulfsolter** [balderdashy/sails#4377]
<!--
======================================================
HELLO, and welcome to the (experimental) Sailsbot
issue-submission system.  If you encounter any
problems with this system, please contact us directly
at http://sailsjs.com/contact.

FIRST THINGS FIRST: before posting, please carefully
read the contribution guide (http://bit.ly/sails-issue-guide),
particularly if this is a security issue or an issue of a
non-technical nature (Github org issues, code of conduct
issues, etc.).  Those should be sent directly at
http://sailsjs.com/contact.

NOW THEN: please fill out the following info
(Don't change text between **)
======================================================
### BEGIN VERSION INFO ###
-->

**Sails version**: 1.0.1
**Node version**: 8.11.1
**NPM version**: 5.6.0
**DB adapter name**: sails-postgresql
**DB adapter version**: 1.0.1
**Operating system**: Ubuntu 16.04.4 LTS

<!--
### END VERSION INFO ###
======================================================
GREAT! - Now read carefully
(Sailsbot will know if you don't)!
======================================================

Before you submit an issue for Sails, please make sure you have read the issue contribution guide (http://bit.ly/sails-issue-guide) carefully, and then verify the following by typing an X in between each set of [ ] brackets below (don't make any other changes to the text!):

### BEGIN PLEDGE ###
- [X] I am experiencing a concrete technical issue (aka a bug) with Sails (ideas and feature proposals should follow the guide for proposing features and enhancements (http://bit.ly/sails-feature-guide), which involves making a pull request).  If you're not 100% certain whether it's a bug or not, that's okay--you may continue.  The worst that can happen is that the issue will be closed and we'll point you in the right direction.

- [X] I am not asking a question about how to use Sails or about whether or not Sails has a certain feature (please refer to the documentation(http://sailsjs.com), or post on http://stackoverflow.com, our Google Group (http://bit.ly/sails-google-group) or our live chat (https://gitter.im/balderdashy/sails).

- [X] I have already searched for related issues, and found none open (if you found a related _closed_ issue, please link to it in your post).

- [X] My issue title is concise, on-topic and polite ("jst.js being removed from layout.ejs on lift" is good; "templates dont work" or "why is sails dumb" are not so good).

- [X] I have tried all the following (if relevant) and my issue remains:
  - Make sure you have the right app lifted.
  - Make sure you've killed the Sails server with CTRL+C and started it again.
  - Make sure you closed any open browser tabs pointed at localhost before starting Sails.
  - Make sure you do not have any other Sails apps running in other terminal windows.
  - Make sure the app you are using to reproduce the issue has a clean node_modules/ directory, meaning:
    * no dependencies are linked (e.g. you haven't run npm link foo)
    * that you haven't made any inline changes to files in the node_modules/ folder
    * that you don't have any weird global dependency loops. The easiest way to double-check any of the above, if you aren't sure, is to run: rm -rf node_modules && npm cache clear && npm install.

and finally...
- [X] **I can provide steps to reproduce this issue that others can follow.**

Ideally, this involves creating a new repo that demonstrates the problem (see instructions at http://bit.ly/sails-issue-repro).  Even though your issue may seem so simple to reproduce that a new repo is unnecessary, you'd be surprised how many solutions present themselves when you start from `sails new` and attempt to recreate your issue from scratch in a new app.  This ensures that the real issue isn't in your user code (a forgotten policy file, perhaps?) or in a third-party module.  If you're absolutely _convinced_ that a new repo is unnecessary, provide clear, concise and _specific_ steps to reproduce the problem in your post (not "create a model then do blueprint create").
### END PLEDGE ###

OKAY--THANKS FOR READING!  Continue posting details of your issue below. -->
<hr/>

Since a recent upgrade (last few weeks, ~March/April 2018) we've lost the ability to see the stack trace when an error is thrown after a model is accessed. When an error is thrown, the local stack trace is not logged but instead that of the last called `sails-postgresql` method.

Example:
````
WARNING: Something seems to be wrong with this function.
It is trying to signal that it has finished AGAIN, after
already resolving/rejecting once.
(silently ignoring this...)

To assist you in hunting this down, here is a stack trace:
```
    at Object.releaseConnection (/home/wherewolf/stack-swallow-test/node_modules/sails-postgresql/helpers/private/connection/release-connection.js:27:6)
    at runQueryCb (/home/wherewolf/stack-swallow-test/node_modules/sails-postgresql/helpers/select.js:166:28)
    at Object.success (/home/wherewolf/stack-swallow-test/node_modules/sails-postgresql/helpers/private/query/run-query.js:114:14)
    at /home/wherewolf/stack-swallow-test/node_modules/machine/lib/private/help-build-machine.js:1354:30
    at _tryToRunCb (/home/wherewolf/stack-swallow-test/node_modules/parley/lib/private/Deferred.js:520:18)
    at /home/wherewolf/stack-swallow-test/node_modules/parley/lib/private/Deferred.js:755:16
    at /home/wherewolf/stack-swallow-test/node_modules/machine/lib/private/help-build-machine.js:966:24
    at Function.handlerCbs.success (/home/wherewolf/stack-swallow-test/node_modules/machine/lib/private/help-build-machine.js:826:26)
    at Query.query (/home/wherewolf/stack-swallow-test/node_modules/machinepack-postgresql/machines/send-native-query.js:128:20)
    at Query.handleReadyForQuery (/home/wherewolf/stack-swallow-test/node_modules/pg/lib/query.js:114:10)
    at Connection.<anonymous> (/home/wherewolf/stack-swallow-test/node_modules/pg/lib/client.js:172:19)
    at emitOne (events.js:116:13)
    at Connection.emit (events.js:211:7)
    at Socket.<anonymous> (/home/wherewolf/stack-swallow-test/node_modules/pg/lib/connection.js:121:12)
    at emitOne (events.js:116:13)
    at Socket.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:250:11)
    at Socket.Readable.push (_stream_readable.js:208:10)
    at TCP.onread (net.js:607:20)
```

 [?] For more help, visit https://sailsjs.com/support
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
````

I've created a replication repo at https://github.com/wulfsolter/stack-swallow-test with issue triggered by running `$ npm start` || `$ npm test`. DB is public and hosted on ElephantSQL so no post-install setup required.

What should happen - When throwing an error inside a controller/service, a stack trace is shown.

````
  good: function (req, res) {
    res.send({ message: 'This logs a stack trace' });
    throw new Error('foo');
  },
````
produces
````
error: Sending 500 ("Server Error") response:                                                                                                 
 Error: foo                                                                                                                                   
    at Object.good [as bar/good] (/home/wherewolf/stack-swallow-test/api/controllers/BarController.js:21:11)                                  
    at /home/wherewolf/stack-swallow-test/node_modules/sails/lib/router/bind.js:247:46                                                        
    at routeTargetFnWrapper (/home/wherewolf/stack-swallow-test/node_modules/sails/lib/router/bind.js:391:9)                                  
    at Layer.handle [as handle_request] (/home/wherewolf/stack-swallow-test/node_modules/express/lib/router/layer.js:95:5)                    
    at next (/home/wherewolf/stack-swallow-test/node_modules/express/lib/router/route.js:137:13)                                              
<< shortened for brevity>>
````

while
````
  bad: function (req, res) {
    Foo
    .find({})
    .exec((error, data) => {
      res.send({ message: 'This does NOT log a meaningful stack trace', data: data });
      throw new Error('foo');
    });
  },
````
produces
````
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -                                                                                                                        
WARNING: Something seems to be wrong with this function.                                                                                                                           
It is trying to signal that it has finished AGAIN, after                                                                                                                           
already resolving/rejecting once.                                                                                                                                                  
(silently ignoring this...)                                                                                                                                                        
                                                                                                                                                                                   
To assist you in hunting this down, here is a stack trace:                                                                                                                         
```                                                                                                                                                                                
    at Object.releaseConnection (/home/wherewolf/stack-swallow-test/node_modules/sails-postgresql/helpers/private/connection/release-connection.js:27:6)                           
    at runQueryCb (/home/wherewolf/stack-swallow-test/node_modules/sails-postgresql/helpers/select.js:166:28)                                                                      
    at Object.success (/home/wherewolf/stack-swallow-test/node_modules/sails-postgresql/helpers/private/query/run-query.js:114:14)                                                 
    at /home/wherewolf/stack-swallow-test/node_modules/machine/lib/private/help-build-machine.js:1354:30                                                                           
    at _tryToRunCb (/home/wherewolf/stack-swallow-test/node_modules/parley/lib/private/Deferred.js:520:18)                                                                         
    at /home/wherewolf/stack-swallow-test/node_modules/parley/lib/private/Deferred.js:755:16                                                                                       
    at /home/wherewolf/stack-swallow-test/node_modules/machine/lib/private/help-build-machine.js:966:24                                                                            
    at Function.handlerCbs.success (/home/wherewolf/stack-swallow-test/node_modules/machine/lib/private/help-build-machine.js:826:26)                                              
    at Query.query (/home/wherewolf/stack-swallow-test/node_modules/machinepack-postgresql/machines/send-native-query.js:128:20)                                                   
    at Query.handleReadyForQuery (/home/wherewolf/stack-swallow-test/node_modules/pg/lib/query.js:114:10)                                                                          
    at Connection.<anonymous> (/home/wherewolf/stack-swallow-test/node_modules/pg/lib/client.js:172:19)                                                                            
    at emitOne (events.js:116:13)                                                                                                                                                  
    at Connection.emit (events.js:211:7)                                                                                                                                           
    at Socket.<anonymous> (/home/wherewolf/stack-swallow-test/node_modules/pg/lib/connection.js:121:12)                                                                            
    at emitOne (events.js:116:13)                                                                                                                                                  
    at Socket.emit (events.js:211:7)                                                                                                                                               
    at addChunk (_stream_readable.js:263:12)                                                                                                                                       
    at readableAddChunk (_stream_readable.js:250:11)                                                                                                                               
    at Socket.Readable.push (_stream_readable.js:208:10)                                                                                                                           
    at TCP.onread (net.js:607:20)                                                                                                                                                  
```                                                                                                                                                                                
                                                                                                                                                                                   
 [?] For more help, visit https://sailsjs.com/support                                                                                                                              
- - - - - - - - - - - - - - - - - - - - - - - - - - - - - -                                                                                                                        
````
which is a stack trace of `sails-postgresql` and not of the controller that threw the error.

Other related GitHub issues (or where terrible stack traces are posted) are https://github.com/balderdashy/sails/issues/4339#issuecomment-382346199 & https://github.com/balderdashy/sails/issues/4312#issuecomment-381683242

**Comment by sailsbot** [comment]
@wulfsolter Thanks for posting, we'll take a look as soon as possible.

---

For help with questions about Sails, [click here](http://sailsjs.com/support).  If you&rsquo;re interested in hiring @sailsbot and her minions in Austin, [click here](http://sailsjs.com/studio).

**Comment by wulfsolter** [comment]
Notes as I debug.....

Passing a second param to `.exec()` as per [Parley](https://github.com/mikermcneil/parley) [README.md](https://github.com/mikermcneil/flaverr/blob/master/README.md)

````
> **EXPERIMENTAL:** As of parley 2.3.x, there is a new, experimental feature that allows you to
> easily provide an extra layer of protection: an optional 2nd argument to `.exec()`.  If specified,
> this function will be used as an uncaught exception handler-- a simple fallback just in case something
> happens to go wrong in your callback function.
>
> This allows you to safely write code like the following without crashing the server:
>
> ```javascript
> User.create({ username: 'foo' }).exec(function (err, result) {
>   if (err) {
>     if (err.code === 'E_UNIQUE') { return res.badRequest('Username already in use.'); }
>     else { return res.serverError(err); }
>   }
>
>   var thisWillNeverWork = JSON.parse('who0ps"thisis totally not valid js{}n');
>
>   return res.json(result);
>
> }, res.serverError);
> ```
>
> Of course, it's still best to be explicit about error handling whenever possible.
> The extra layer of protection is just that-- it's here to help prevent issues
> stemming from the myriad runtime edge cases it's almost impossible to anticipate
> when building a production-ready web application.
````

allows logging of trace like so

````
  bad: function (req, res) {
    Foo
    .find({})
    .exec((error, data) => {
      res.send({ message: 'This does NOT log a meaningful stack trace', data: data });
      throw new Error('foo');
    }, function (trace) {
      // this shows more meaningful trace
      console.log(trace);
    });
  },
````
which includes ~14 frames on top of `Object.releaseConnection()`

**Comment by wulfsolter** [comment]
Interim work-around is to put a `console.log(err.raw);` just above the console.warn() line here 

https://github.com/mikermcneil/parley/blob/61fabbaa81257e9c38a782a11a2c41635d4af5df/lib/private/Deferred.js#L683-L689

**Comment by wulfsolter** [comment]
After stepping through [Parley]( https://github.com/mikermcneil/parley) and [Flaverr](https://github.com/mikermcneil/flaverr), I don't feel confident enough to see where @mikermcneil is headed. I trust that swallowing the top dozen frames of a stack trace are not the intention, but I'm not sure where to adjust things to remain true to the direction for Parley/Flaverr so am a touch hesitant to submit a PR.

**Comment by mikermcneil** [comment]
@wulfsolter try setting `DEBUG=true` (env var)

**Comment by mikermcneil** [comment]
@wulfsolter nvm just noticed you customized the npm start script- will take a closer look

**Comment by mikermcneil** [comment]
@wulfsolter https://github.com/wulfsolter/stack-swallow-test/pull/1

**Comment by wulfsolter** [comment]
@johnabrams7 PR on https://github.com/mikermcneil/parley/pull/6 would close this

**Comment by mikermcneil** [comment]
published in parley@3.8.3, which will be picked up automatically in a Sails app upon next `rm -rf package-lock.json node_modules && npm install`

(may or may not need all the smashing and crashing above, but when it comes to npm CLI, a little smashing and crashing usually doesn't hurt)

Thanks @wulfsolter for getting to the bottom of this!

**Comment by netrevisanto** [comment]
Thanks a lot, was really annoyed by that.
